hi-ml-cpath/src/health_cpath/preprocessing/loading.py

```diff
@@ -71,49 +138,222 @@ class LoadROId(MapTransform):
-        scale = slide_obj.resolutions['level_downsamples'][highest_level]
-        slide = load_slide_at_level(self.reader, slide_obj, level=highest_level)
+    def _load_slide_at_level(self, slide_obj: _OpenSlideOrCuImage, level: int) -> np.ndarray:
+        """Load full slide array at the given magnification level.
+
+        This is a manual workaround for a MONAI bug (https://github.com/Project-MONAI/MONAI/issues/3415)
+        fixed in a currently unreleased PR (https://github.com/Project-MONAI/MONAI/pull/3417).
+
+        :param reader: A MONAI `WSIReader` using cuCIM backend.
+        :param slide_obj: The cuCIM image object returned by `reader.read(<image_file>)`.
+        :param level: Index of the desired magnification level as defined in the `slide_obj` headers.
+        :return: The loaded image array in (C, H, W) format.
+        """
+        size = self._get_size_at_level(slide_obj, level)
+        slide, _ = self.reader.get_data(slide_obj, size=size, level=level)  # loaded as RGB PIL image
+        return slide
+
+    def _get_bounding_box(self, slide_obj: _OpenSlideOrCuImage) -> box_utils.Box:
+        """Estimate bounding box at the lowest resolution (i.e. highest level) of the slide."""
+        highest_level = self._get_highest_level(slide_obj)
+        scale = self._get_scale_at_level(slide_obj, highest_level)
+        slide = self._load_slide_at_level(slide_obj, level=highest_level)
 
         foreground_mask, threshold = segment_foreground(slide, self.foreground_threshold)
+        self.foreground_threshold = threshold
         bbox = scale * box_utils.get_bounding_box(foreground_mask).add_margin(self.margin)
-        return bbox, threshold
+        return bbox
 
     def __call__(self, data: Dict) -> Dict:
-        image_obj: CuImage = self.reader.read(data[self.image_key])
+        image_obj: _OpenSlideOrCuImage = self.reader.read(data[self.image_key])
 
-        level0_bbox, threshold = self._get_bounding_box(image_obj)
+        level0_bbox = self._get_bounding_box(image_obj)
 
         # cuCIM/OpenSlide takes absolute location coordinates in the level 0 reference frame,
         # but relative region size in pixels at the chosen level
         origin = (level0_bbox.y, level0_bbox.x)
-        scale = image_obj.resolutions['level_downsamples'][self.level]
+        scale = self._get_scale_at_level(image_obj, self.level)
         scaled_bbox = level0_bbox / scale
 
         data[self.image_key], _ = self.reader.get_data(image_obj, location=origin, level=self.level,
                                                        size=(scaled_bbox.h, scaled_bbox.w))
         data[SlideKey.ORIGIN] = origin
         data[SlideKey.SCALE] = scale
-        data[SlideKey.FOREGROUND_THRESHOLD] = threshold
+        data[SlideKey.FOREGROUND_THRESHOLD] = self.foreground_threshold
 
         image_obj.close()
         return data
+
+
+class LoadMaskROId(MapTransform, BaseLoadROId, Generic[_OpenSlideOrCuImage]):
```

Comment:
Please make sure I haven't accidentally messed up something while moving this transform to this file. original in panda_dataset.